libsolidity/ast/AST.cpp

```diff
@@ -480,6 +480,63 @@ bool VariableDeclaration::canHaveAutoType() const
 	return isLocalVariable() && !isCallableParameter();
 }
 
+bool VariableDeclaration::isEventParameter() const
+{
+	return dynamic_cast<EventDefinition const*>(scope()) != nullptr;
+}
+
+bool VariableDeclaration::isReferenceType() const
+{
+	solAssert(typeName()->annotation().type, "Can only be called after reference resolution");
+	return dynamic_cast<ReferenceType const*>(typeName()->annotation().type.get());
+}
+
+set<VariableDeclaration::Location> VariableDeclaration::allowedStorageLocations() const
+{
+	using Location = VariableDeclaration::Location;
+
+	if (!isReferenceType())
+		return set<Location>{ Location::Default };
+
+	if (isEventParameter())
+		return set<Location>{ Location::Default };
+	else if (isConstant())
+		return set<Location>{Location::Default, Location::Memory};
+	else if (isExternalCallableParameter())
+	{
+		auto const& contract = dynamic_cast<ContractDefinition const&>(
+				*dynamic_cast<Declaration const&>(*scope()).scope()
+		);
+		if (contract.isLibrary())
+			return set<Location>{ Location::CallData, Location::Default, Location::Storage };
+		else
+			return set<Location>{ Location::CallData, Location::Default };
+	}
+	else if (isCallableParameter())
+	{
+		auto const& varScope = dynamic_cast<Declaration const&>(*scope());
+		auto const& contract = dynamic_cast<ContractDefinition const&>(
+				*dynamic_cast<Declaration const&>(*scope()).scope()
+		);
+		if (varScope.isPublic())  // Public function parameter.
+		{
+			if (contract.isLibrary())
+				return set<Location>{ Location::Memory, Location::Storage };
+			else
+				return set<Location>{ Location::Memory };
+		}
+		else  // Internal or private function parameter.
+			return set<Location>{ Location::Memory, Location::Storage };
+	}
+	else  // Variable. TODO: add DataLocation::Calldata once implemented for local variables.
+	{
+		if (!isLocalVariable())
+			return set<Location>{ Location::Memory, Location::Storage, Location::Default };
```

Comment:
Please remind me, why do we allow `memory` and `storage` in this case?

Also: Please swap the two cases to simplify the condition (no negation is easier to understand).

Reply:
The case of not a local variable is used for struct members, I have added a more explicit way of checking for this.

Reply:
@chriseth Also, I'm running into small issues with function type variables. Their parameter and return parameter variables' scopes are going to whatever scope the actual function type is. For example, in `struct S { function() returns (S storage)[] x; }` if we call `isReturnParameter()`, or `isCallableParameter()`, etc. on the return paramter of the function, we get false. However, if we run `isStructMember()`, which is a function I added that will be in the next commit (returns `dynamic_cast<StructDefinition const*>(scope()) != nullptr;`) we get true. Should parameters of function type variables behave differently than normal function parameters?

Reply:
I think the solution would be to modify `scope()` a little. All points where `scope()` is called have to be changed to also add the possibility that `scope()` returns a `FunctionTypeName` and not a `CallableDeclaration` These include:


 - `Scopable::sourceUnit()`
 - `VariableDeclaration::isLocalVariable()`
 - `VariableDeclaration::isCallableParameter()`
 - `VariableDeclaration::isReturnParameter()`
 - `VariableDeclaration::isExternalCallableParameter()`

So in essence, FunctionTypeName parameters should be handled in the same way as struct elements should be handled. This for example also concerns `DeclarationRegistrationHelper::registerDeclaration`.

You should also add a visit-overload for `FunctionTypeName` into `DeclarationRegistrationHelper` in the same way as it is done for StructDefinition.